strawberryfields/devicespecs/device_specs.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Abstract base class for storing device data for validation"""
+from typing import List, Set, Dict
+import abc
+
+import blackbird
+from blackbird.utils import to_DiGraph
+
+
+class DeviceSpecs(abc.ABC):
+    """Abstract base class for backend data"""
+
+    @property
+    @abc.abstractmethod
+    def modes(self) -> int:
+        """The supported number of modes of the device.
+
+        If the device supports arbitrary number of modes, set this to 0.
+
+        Returns:
+            int: number of supported modes
+        """
+
+    @property
+    @abc.abstractmethod
+    def remote(self) -> bool:
```

Comment:
I suppose we also want to have a `local` property as well then, since various combinations or remote/local execution are possible depending on the device

Reply:
Ah, I didn't think of this. How would this look? I assumed `remote` and `local` would be mutually exclusive

Reply:
There are two options:

- Have `DeviceSpec`s registered separately for local and remote versions of backends that are otherwise the same (e.g., Gaussian backend running locally or remotely). 

- Have a single `DeviceSpec` for a backend type, and extra properties `local` and `remote` which indicate what operation types are supported (could be one or the other, or both).

I think second is much easier

Reply:
Added the second option :+1: